upstairs/src/volume.rs

```diff
@@ -1092,15 +1094,355 @@ impl Volume {
+        let n_opts = &n_sub_volume.opts;
+        if o_opts.id != n_opts.id {
+            crucible_bail!(
+                ReplaceRequestInvalid,
+                "sub_volume opts id invalid {:?} vs. {:?}",
+                o_opts.id,
+                n_opts.id
+            )
+        }
+        if o_opts.lossy != n_opts.lossy {
+            crucible_bail!(
+                ReplaceRequestInvalid,
+                "sub_volume opts lossy invalid {:?} vs. {:?}",
+                o_opts.lossy,
+                n_opts.lossy
+            )
+        }
+        if o_opts.flush_timeout != n_opts.flush_timeout {
+            crucible_bail!(
+                ReplaceRequestInvalid,
+                "sub_volume opts flush_timeout mismatch {:?} vs. {:?}",
+                o_opts.flush_timeout,
+                n_opts.flush_timeout
+            )
+        }
+        if o_opts.key != n_opts.key {
+            crucible_bail!(
+                ReplaceRequestInvalid,
+                "sub_volume opts key invalid {:?} vs. {:?}",
+                o_opts.key,
+                n_opts.key
+            )
+        }
+        if o_opts.cert_pem != n_opts.cert_pem {
+            crucible_bail!(
+                ReplaceRequestInvalid,
+                "sub_volume opts cert_pem invalid {:?} vs. {:?}",
+                o_opts.cert_pem,
+                n_opts.cert_pem
+            )
+        }
+        if o_opts.key_pem != n_opts.key_pem {
+            crucible_bail!(
+                ReplaceRequestInvalid,
+                "sub_volume opts key_pem invalid {:?} vs. {:?}",
+                o_opts.key_pem,
+                n_opts.key_pem
+            )
+        }
+        if o_opts.root_cert_pem != n_opts.root_cert_pem {
+            crucible_bail!(
+                ReplaceRequestInvalid,
+                "sub_volume opts root_cert_pem invalid {:?} vs. {:?}",
+                o_opts.root_cert_pem,
+                n_opts.root_cert_pem
+            )
+        }
+        if o_opts.control != n_opts.control {
+            crucible_bail!(
+                ReplaceRequestInvalid,
+                "sub_volume opts control invalid {:?} vs. {:?}",
+                o_opts.control,
+                n_opts.control
+            )
+        }
+        if o_opts.read_only != n_opts.read_only {
+            crucible_bail!(
+                ReplaceRequestInvalid,
+                "sub_volume opts read_only invalid {:?} vs. {:?}",
+                o_opts.read_only,
+                n_opts.read_only
+            )
+        }
+        if o_opts.target == n_opts.target {
+            crucible_bail!(
+                ReplaceRequestInvalid,
+                "sub_volume opts target unchanged {:?} vs. {:?}",
+                o_opts.target,
+                n_opts.target
+            )
+        }
+
+        // Walk the targets, we can only have one different
+        let mut new_target_cid = None;
+        for cid in 0..3 {
+            if o_opts.target[cid] != n_opts.target[cid] {
```

Comment:
Does target ordering matter in other contexts? (I'm fretting about a hypothetical future bug where Nexus replaces a target by calling `remove()` to take one out and `push()` to put one in...)

Reply:
Yes and no.  The initial order for targets does not matter, but once we have started the upstairs, whomever is at index 0 must remain their for the life of that upstairs process.  We can't do `remove()` and `push()` for that reason inside Crucible upstairs.

The upstairs is also pretty dependent on three targets always existing (even if it can't connect to one of them).

If you were to stop the upstairs and create a new upstairs, then the order can be different and the upstairs will be just fine.  There is no dependency on a specific downstairs being at a specific target index when starting cold.

You are correct to fret about Nexus trying to `remove()` and `push()`, and when we write that side of the downstairs replacement code, it will have to do a replacement of the specific target at the index it currently resides in.